sam.c

```diff
@@ -127,26 +128,95 @@ bam_hdr_t *bam_hdr_read(BGZF *fp)
         return 0;
     }
     h = bam_hdr_init();
+    if (!h) return 0;
```

Comment:
Strictly, yes, although as you note the clean-up code is written with the assumption that *h is valid.  Hence it just bails at this point, and with the EOF check above.  It's a fairly easy change to make it print the message, though.